src/R2N.jl

```diff
@@ -0,0 +1,412 @@
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+
+mutable struct R2NSolver{
+  T,
+  V,
+  Op <: AbstractLinearOperator{T},
+  Op2 <: AbstractLinearOperator{T},
+  Sub <: Union{KrylovSolver{T, T, V}, ShiftedLBFGSSolver},
+} <: AbstractOptimizationSolver
+  x::V
+  cx::V
+  gx::V
+  gn::V
+  σ::T
+  μ::T
+  H::Op
+  opI::Op2
+  Hs::V
+  s::V
+  obj_vec::V # used for non-monotone behaviour
+  subsolver_type::Sub
+  cgtol::T
+end
+
+function R2NSolver(
+  nlp::AbstractNLPModel{T, V};
+  non_mono_size = 1,
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+) where {T, V}
+  nvar = nlp.meta.nvar
+  x = V(undef, nvar) # vs similar(nlp.meta.x0)
+  cx = V(undef, nvar)
+  gx = V(undef, nvar)
+  gn = isa(nlp, QuasiNewtonModel) ? V(undef, nvar) : V(undef, 0)
+  Hs = V(undef, nvar)
+  H = isa(nlp, QuasiNewtonModel) ? nlp.op : hess_op!(nlp, x, Hs)
+  opI = opEye(T, nvar)
+  Op = typeof(H)
+  Op2 = typeof(opI)
+  σ = zero(T)
+  μ = zero(T)
+  s = V(undef, nvar)
+  cgtol = one(T) # must be ≤ 1.0
+  obj_vec = fill(typemin(T), non_mono_size)
+  subsolver =
+    isa(subsolver_type, Type{ShiftedLBFGSSolver}) ? subsolver_type() : subsolver_type(nvar, nvar, V)
+
+  Sub = typeof(subsolver)
+  return R2NSolver{T, V, Op, Op2, Sub}(
+    x,
+    cx,
+    gx,
+    gn,
+    σ,
+    μ,
+    H,
+    opI,
+    Hs,
+    s,
+    obj_vec,
+    subsolver,
+    cgtol,
+  )
+end
+
+function SolverCore.reset!(solver::R2NSolver{T}) where {T}
+  fill!(solver.obj_vec, typemin(T))
+  reset!(solver.H)
+  solver
+end
+function SolverCore.reset!(solver::R2NSolver{T}, nlp::AbstractNLPModel) where {T}
+  fill!(solver.obj_vec, typemin(T))
+  @assert (length(solver.gn) == 0) || isa(nlp, QuasiNewtonModel)
+  solver.H = isa(nlp, QuasiNewtonModel) ? nlp.op : hess_op!(nlp, solver.x, solver.Hs)
+  solver
+end
+
+@doc (@doc R2NSolver) function R2N(
+  nlp::AbstractNLPModel{T, V};
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+  non_mono_size = 1,
+  kwargs...,
+) where {T, V}
+  solver = R2NSolver(nlp; non_mono_size = non_mono_size, subsolver_type = subsolver_type)
+  return solve!(solver, nlp; non_mono_size = non_mono_size, kwargs...)
+end
+
+function SolverCore.solve!(
+  solver::R2NSolver{T, V},
+  nlp::AbstractNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = nlp.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  η1 = T(0.0001),
+  η2 = T(0.001),
+  λ = T(2),
+  σmin = zero(T),
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  max_iter::Int = typemax(Int),
+  verbose::Int = 0,
+  subsolver_verbose::Int = 0,
+  non_mono_size = 1,
+) where {T, V}
+  unconstrained(nlp) || error("R2N should only be called on unconstrained problems.")
+  if non_mono_size < 1
+    error("non_mono_size must be greater than or equal to 1")
+  end
+  if (solver.subsolver_type isa ShiftedLBFGSSolver && !isa(nlp, LBFGSModel))
```

Comment:
This could happen when you construct the solver. See comments above.

Reply:
so 2 new tests:
1. non_mono_size <1 and subsolver type ?